### DISCUSSION TITLE
Left recursion vs right recursion: does it matter?

### DISCUSSION THREAD

**Discussion by alemuller** [tree-sitter/tree-sitter#1543] [Q&A] [answered]
From bison doc:

> Any kind of sequence can be defined using either left recursion or right recursion, but you should always use left recursion, because it can parse a sequence of any number of elements with bounded stack space

Does this apply for tree-sitter? && For `repeat`?

Is writing lists in this format any better?
```js
list: $ => seq(
  repeat1(seq($.elem, SEP)),
  $.elem
),
```

**Comment by maxbrunsfeld** [comment, marked as answer]
With Tree-sitter, you should always use ‘repeat’ or ‘repeat1’, as opposed to writing recursive rules, if you expect there to be many repetitions.

Tree-sitter will represent the repetition as a balanced binary tree, instead of a leftward or rightward leaning linked list, which is basically what you’d get with explicit recursion. The balancing gives much better performance when traversing or editing the tree.

**Reply by alemuller** [reply]
Interesting.

A follow up question.

With the assumption that `repeat` has (usually) better performance when compared with recursion.

Is is reasonable to expect that writing *expressions* with `repeat` to be faster than how they are traditionally done (with left recursion on `$._expression`)?

Putting this into code:

1. *Traditional way*:
```js
_expr: $ => choice(
  $.foo,
  $.bar,
  $.primary
),

foo: $ => prec.left(1,seq(
  $._expr, OP1, $._expr
)),

bar: $ => prec.left(2,seq(
  $._expr, OP2, $._expr
)),
```

2. *Alternative way*:
```js
_expr: $ => choice(
  $.foo,
  $.bar,
  $.primary
),

foo: $ => seq(
  // This has the same effect as prec.left, right?
  repeat(seq($._foo_expr, OP1)), $._foo_expr
),

_foo_expr: $ => choice(
  $.bar,
  $.primary
),

bar: $ => seq(
  repeat(seq($._bar_expr, OP2)), $._bar_expr
),

_bar_expr: $ => choice(
  $.primary
),
```
(Haven't tried to run them, might be some typos).

Putting aside the AST differences and the extra verboseness. Can I say it is "better" to write expression in the *alternative way*?

Is this the reason why you favored writing expression with a single `$.binary_expression` instead of many rules?

(If I saying something that doesn't make sense or using wrong terminology, please point out :P)

**Reply by maxbrunsfeld** [reply]
I think it's only important to use `repeat` if you expect there to be an arbitrarily long list of things. For things like `binary_expression`, I think it's fine to use recursion.
